python/nx-cugraph/nx_cugraph/algorithms/components/strongly_connected.py

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import networkx as nx
+import pylibcugraph as plc
+
+from nx_cugraph.convert import _to_directed_graph
+from nx_cugraph.utils import (
+    _groupby,
+    index_dtype,
+    networkx_algorithm,
+    not_implemented_for,
+)
+
+__all__ = [
+    "number_strongly_connected_components",
+    "strongly_connected_components",
+    "is_strongly_connected",
+]
+
+
+def _strongly_connected_components(G):
+    # TODO: create utility function to convert just the indices to CSR
+    # TODO: this uses a legacy PLC function (strongly_connected_components)
+    N = len(G)
+    indices = cp.lexsort(cp.vstack((G.dst_indices, G.src_indices)))
+    dst_indices = G.dst_indices[indices]
+    offsets = cp.searchsorted(
+        G.src_indices, cp.arange(N + 1, dtype=index_dtype), sorter=indices
+    ).astype(index_dtype)
+    labels = cp.zeros(N, dtype=index_dtype)
+    plc.strongly_connected_components(
+        offsets=offsets,
+        indices=dst_indices,
+        weights=None,
+        num_verts=N,
+        num_edges=dst_indices.size,
+        labels=labels,
+    )
+    return labels
+
+
+@not_implemented_for("undirected")
+@networkx_algorithm(version_added="24.02", plc="strongly_connected_components")
+def strongly_connected_components(G):
+    G = _to_directed_graph(G)
+    if G.src_indices.size == 0:
+        return [{key} for key in G._nodeiter_to_iter(range(len(G)))]
+    labels = _strongly_connected_components(G)
+    groups = _groupby(labels, cp.arange(len(G), dtype=index_dtype))
+    return (G._nodearray_to_set(connected_ids) for connected_ids in groups.values())
+
+
+@not_implemented_for("undirected")
+@networkx_algorithm(version_added="24.02", plc="strongly_connected_components")
+def number_strongly_connected_components(G):
+    G = _to_directed_graph(G)
+    if G.src_indices.size == 0:
+        return len(G)
```

Comment:
just curious, no action needed: is this expected to always return 0 in this case?  If so, is there a reason calling `len()` is preferred over just returning 0?

Reply:
Maybe I should use `G.number_of_edges()` instead of `G.src_indices.size` (but for some reason the latter is easier for me to remember and reason about). Anyway, if the number of _edges_ are zero, the the number of components is the number of _nodes_, hence we can't simply return 0.

Reply:
Oh I see, `number_of_edges` actually does a lot more work. If we want to know if there are exactly 0 edges, `G.src_indices.size` works great.